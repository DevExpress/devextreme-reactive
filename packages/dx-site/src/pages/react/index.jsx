import * as React from 'react';
import Layout from '../../components/layout';

const IndexPage = () => (
  <Layout>
    React index
  </Layout>
);

export default IndexPage;
