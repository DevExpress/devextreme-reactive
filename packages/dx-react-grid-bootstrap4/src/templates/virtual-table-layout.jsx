import * as React from 'react';
import {
  TableLayout,
  VirtualTableLayout as VirtualTableLayoutCore,
} from '@devexpress/dx-react-grid';

const MINIMAL_COLUMN_WIDTH = 150;

export const VirtualTableLayout = props => (
  <TableLayout
    layoutComponent={VirtualTableLayoutCore}
    minColumnWidth={MINIMAL_COLUMN_WIDTH}
    {...props}
  />
);
