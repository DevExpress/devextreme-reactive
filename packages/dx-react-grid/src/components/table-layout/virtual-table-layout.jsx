import * as React from 'react';
import { findDOMNode } from 'react-dom';
import * as PropTypes from 'prop-types';
import { ColumnGroup } from './column-group';
import { RowLayout } from './row-layout';
import {
  getVisibleRows,
  firstVisibleRowOffset,
} from './virtual-table-utils';

export class VirtualTableLayout extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      rowHeights: new Map(),
      viewportTop: 0,
    };

    this.rowRefs = new Map();
    this.updateViewport = this.updateViewport.bind(this);
    this.registerRowRef = this.registerRowRef.bind(this);
    this.getRowHeight = this.getRowHeight.bind(this);
  }
  componentDidMount() {
    this.storeRowHeights();
  }
  componentWillReceiveProps(nextProps) {
    if (
      this.props.headerRows !== nextProps.headerRows ||
      this.props.rows !== nextProps.rows
    ) {
      const { rowHeights: prevRowHeight } = this.state;
      this.setState({
        rowHeights: [...nextProps.headerRows, ...nextProps.rows].reduce(
          (acc, row) => {
            const rowHeight = prevRowHeight.get(row.key);
            if (rowHeight !== undefined) {
              acc.set(row.key, rowHeight);
            }
            return acc;
          },
          new Map(),
        ),
      });
    }
  }
  componentDidUpdate() {
    this.storeRowHeights();
  }
  getRowHeight(row) {
    const storedHeight = this.state.rowHeights.get(row.key);
    if (storedHeight !== undefined) return storedHeight;
    if (row.height) return row.height;
    return this.props.estimatedRowHeight;
  }
  getVisibleRows({ rows, headerRows, height } = this.props, top = this.state.viewportTop) {
    const headHeight = headerRows.reduce((acc, row) => acc + this.getRowHeight(row), 0);
    return getVisibleRows(rows, top, height - headHeight, this.getRowHeight);
  }
  storeRowHeights() {
    const rowsWithChangedHeights = Array.from(this.rowRefs.entries())
      // eslint-disable-next-line react/no-find-dom-node
      .map(([row, ref]) => [row, findDOMNode(ref).getBoundingClientRect().height])
      .filter(([row, height]) => height !== this.getRowHeight(row));

    if (rowsWithChangedHeights.length) {
      const prevVisibleBodyRows = this.getVisibleRows();

      const { rowHeights } = this.state;
      rowsWithChangedHeights
        .forEach(([row, height]) => rowHeights.set(row.key, height));

      this.setState({
        rowHeights,
      });

      const visibleBodyRows = this.getVisibleRows();
      const scrollOffset = firstVisibleRowOffset(prevVisibleBodyRows, visibleBodyRows);
      if (scrollOffset !== 0) {
        // eslint-disable-next-line react/no-find-dom-node
        findDOMNode(this).scrollTop += scrollOffset;
      }
    }
  }
  registerRowRef(row, ref) {
    if (ref === null) {
      this.rowRefs.delete(row);
    } else {
      this.rowRefs.set(row, ref);
    }
  }
  updateViewport(e) {
    const node = e.target;

    if (node !== e.currentTarget) {
      return;
    }

    // NOTE: prevent iOS to flicker in bounces
    if (node.scrollTop < 0 ||
      node.scrollLeft < 0 ||
      node.scrollLeft + node.clientWidth > node.scrollWidth ||
      node.scrollTop + node.clientHeight > node.scrollHeight) {
      return;
    }

    if (this.viewportTop !== node.scrollTop) {
      this.setState({
        viewportTop: node.scrollTop,
      });
    }
  }
  render() {
    const {
      headerRows, columns,
      minWidth, height,
      containerComponent: Container,
      headTableComponent: HeadTable,
      tableComponent: Table,
      headComponent: Head,
      bodyComponent: Body,
      rowComponent, cellComponent,
    } = this.props;
    const visibleBodyRows = this.getVisibleRows();

    return (
      <Container
        style={{ height: `${height}px` }}
        onScroll={this.updateViewport}
      >
        {!!headerRows.length && (
          <HeadTable
            style={{ minWidth: `${minWidth}px` }}
          >
            <ColumnGroup columns={columns} />
            <Head>
              {headerRows.map(row => (
                <RowLayout
                  key={row.key}
                  ref={ref => this.registerRowRef(row, ref)}
                  row={row}
                  columns={columns}
                  rowComponent={rowComponent}
                  cellComponent={cellComponent}
                />
              ))}
            </Head>
          </HeadTable>
        )}
        <Table
          style={{ minWidth: `${minWidth}px` }}
        >
          <ColumnGroup columns={columns} />
          <Body>
            {visibleBodyRows.map((visibleRow) => {
              if (visibleRow.type === 'stub') {
                return (
                  <tr key={visibleRow.key} style={{ height: `${visibleRow.height}px` }} />
                );
              }
              const { row } = visibleRow;
              return (
                <RowLayout
                  key={row.key}
                  ref={ref => this.registerRowRef(row, ref)}
                  row={row}
                  columns={columns}
                  rowComponent={rowComponent}
                  cellComponent={cellComponent}
                />
              );
            })}
          </Body>
        </Table>
      </Container>
    );
  }
}

VirtualTableLayout.propTypes = {
  minWidth: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  headerRows: PropTypes.array,
  rows: PropTypes.array.isRequired,
  columns: PropTypes.array.isRequired,
  cellComponent: PropTypes.func.isRequired,
  rowComponent: PropTypes.func.isRequired,
  bodyComponent: PropTypes.func.isRequired,
  headComponent: PropTypes.func,
  tableComponent: PropTypes.func.isRequired,
  headTableComponent: PropTypes.func,
  containerComponent: PropTypes.func.isRequired,
  estimatedRowHeight: PropTypes.number.isRequired,
};

VirtualTableLayout.defaultProps = {
  headerRows: [],
  headComponent: () => null,
  headTableComponent: () => null,
};
