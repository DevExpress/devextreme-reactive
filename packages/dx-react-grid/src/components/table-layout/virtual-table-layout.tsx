import * as React from 'react';
import { MemoizedFunction, memoize } from '@devexpress/dx-core';
import {
  TableColumn, GetColumnWidthFn, getCollapsedGrids,
  getColumnWidthGetter, TABLE_STUB_TYPE, getViewport, GridViewport,
} from '@devexpress/dx-grid-core';
import { VirtualTableLayoutState, VirtualTableLayoutProps } from '../../types';
import { findDOMNode } from 'react-dom';
import { VirtualTableLayoutBlock } from './virtual-table-layout-block';
import { Sizer } from '@devexpress/dx-react-core';

const AUTO_HEIGHT = 'auto';

const defaultProps = {
  headerRows: [],
  footerRows: [],
  headComponent: () => null,
  headTableComponent: () => null,
  footerComponent: () => null,
  footerTableComponent: () => null,
};
type PropsType = VirtualTableLayoutProps & typeof defaultProps;

/** @internal */
// tslint:disable-next-line: max-line-length
export class VirtualTableLayout extends React.PureComponent<PropsType, VirtualTableLayoutState> {
  static defaultProps = defaultProps;
  getColumnWidthGetter: MemoizedFunction<[TableColumn[], number, number], GetColumnWidthFn>;
  rowRefs = new Map();
  blockRefs = new Map();
  viewportTop = 0;
  containerHeight = 600;
  containerWidth = 800;
  viewportLeft = 0;

  constructor(props) {
    super(props);

    this.state = {
      rowHeights: new Map<any, number>(),
      height: 0,
      headerHeight: 0,
      bodyHeight: 0,
      footerHeight: 0,
      visibleRowBoundaries: {},
    };

    const headerHeight = props.headerRows
      .reduce((acc, row) => acc + this.getRowHeight(row), 0);
    const footerHeight = props.footerRows
      .reduce((acc, row) => acc + this.getRowHeight(row), 0);

    this.state = {
      ...this.state,
      headerHeight,
      footerHeight,
    };

    this.getColumnWidthGetter = memoize(
      (tableColumns, tableWidth, minColumnWidth) => (
        getColumnWidthGetter(tableColumns, tableWidth, minColumnWidth)
      ),
    );
  }

  componentDidMount() {
    this.storeRowHeights();
    this.storeBlockHeights();
  }

  componentDidUpdate(prevProps) {
    this.storeRowHeights();
    this.storeBlockHeights();

    const { bodyRows, columns } = this.props;

    // NOTE: the boundaries depend not only on scroll position and container dimensions
    // but on body rows too. This boundaries update is especially important when
    // lazy loading is used because by the time that all involved events are handled
    // no rows are loaded yet.
    const bodyRowsChanged = prevProps.bodyRows !== bodyRows;
    // Also it's the only place where we can respond to the column count change
    const columnCountChanged = prevProps.columns.length !== columns.length;

    if (bodyRowsChanged || columnCountChanged) {
      this.updateViewport();
    }
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    const { rowHeights: prevRowHeight } = prevState;
    const rowHeights = [...nextProps.headerRows, ...nextProps.bodyRows, ...nextProps.footerRows]
      .reduce(
        (acc, row) => {
          const rowHeight = prevRowHeight.get(row.key);
          if (rowHeight !== undefined) {
            acc.set(row.key, rowHeight);
          }
          return acc;
        },
        new Map(),
      );
    return { rowHeights };
  }

  getRowHeight = (row) => {
    const { rowHeights } = this.state;
    const { estimatedRowHeight } = this.props;
    if (row) {
      const storedHeight = rowHeights.get(row.key);
      if (storedHeight !== undefined) return storedHeight;
      if (row.height) return row.height;
    }
    return estimatedRowHeight;
  }

  registerRowRef = (row, ref) => {
    if (ref === null) {
      this.rowRefs.delete(row);
    } else {
      this.rowRefs.set(row, ref);
    }
  }

  registerBlockRef = (name, ref) => {
    if (ref === null) {
      this.blockRefs.delete(name);
    } else {
      this.blockRefs.set(name, ref);
    }
  }

  storeRowHeights() {
    const rowsWithChangedHeights = Array.from(this.rowRefs.entries())
      .map(([row, ref]) => [row, findDOMNode(ref)])
      .filter(([, node]) => !!node)
      .map(([row, node]) => [row, node.getBoundingClientRect().height])
      .filter(([row]) => row.type !== TABLE_STUB_TYPE)
      .filter(([row, height]) => height !== this.getRowHeight(row));

    if (rowsWithChangedHeights.length) {
      const { rowHeights } = this.state;
      rowsWithChangedHeights
        .forEach(([row, height]) => rowHeights.set(row.key, height));

      this.setState({
        rowHeights,
      });
    }
  }

  storeBlockHeights() {
    const getBlockHeight = blockName => (this.blockRefs.get(blockName)
      ? (findDOMNode(this.blockRefs.get(blockName)) as HTMLElement).getBoundingClientRect().height
      : 0
    );
    const headerHeight = getBlockHeight('header');
    const bodyHeight = getBlockHeight('body');
    const footerHeight = getBlockHeight('footer');

    const {
      headerHeight: prevHeaderHeight,
      bodyHeight: prevBodyHeight,
      footerHeight: prevFooterHeight,
    } = this.state;

    if (prevHeaderHeight !== headerHeight
      || prevBodyHeight !== bodyHeight
      || prevFooterHeight !== footerHeight) {
      this.setState({
        headerHeight,
        bodyHeight,
        footerHeight,
      });
    }
  }

  onScroll = (e) => {
    const node = e.target;

    if (this.shouldSkipScrollEvent(e)) {
      return;
    }

    const { scrollTop: viewportTop, scrollLeft: viewportLeft } = node;

    this.viewportTop = viewportTop;
    this.viewportLeft = viewportLeft;

    this.updateViewport();
  }

  handleContainerSizeChange = ({ width, height }) => {
    this.containerHeight = height;
    this.containerWidth = width;

    this.updateViewport();
  }

  shouldSkipScrollEvent(e) {
    const node = e.target;

    // NOTE: prevent nested scroll to update viewport
    if (node !== e.currentTarget) {
      return true;
    }
    // NOTE: normalize position:
    // in Firefox and Chrome (zoom > 100%) when scrolled to the bottom
    // in Edge when scrolled to the right edge
    const correction = 1;
    const nodeHorizontalOffset = parseInt(node.scrollLeft + node.clientWidth, 10) - correction;
    const nodeVerticalOffset = parseInt(node.scrollTop + node.clientHeight, 10) - correction;
    // NOTE: prevent iOS to flicker in bounces and correct rendering on high dpi screens
    if (node.scrollTop < 0
      || node.scrollLeft < 0
      || nodeHorizontalOffset > Math.max(node.scrollWidth, node.clientWidth)
      || nodeVerticalOffset > Math.max(node.scrollHeight, node.clientHeight)) {
      return true;
    }

    return false;
  }

  updateViewport() {
    const { viewport, setViewport } = this.props;
    const newViewport = this.calculateViewport();

    if (viewport !== newViewport) {
      setViewport(newViewport);
    }
  }

  calculateViewport() {
    const { state, viewportTop, viewportLeft, containerHeight, containerWidth } = this;
    const {
      loadedRowsStart,
      bodyRows,
      headerRows,
      footerRows,
      estimatedRowHeight,
      columns,
      minColumnWidth,
      isDataRemote,
      viewport,
    } = this.props;
    const getColumnWidth = this.getColumnWidthGetter(columns, containerWidth, minColumnWidth!);

    return getViewport(
      { ...state, viewportTop, viewportLeft, containerHeight, containerWidth },
      { loadedRowsStart, columns, bodyRows, headerRows, footerRows, isDataRemote, viewport },
      estimatedRowHeight, this.getRowHeight, getColumnWidth,
    );
  }

  getCollapsedGrids(viewport: GridViewport) {
    const { containerWidth, viewportLeft } = this;
    const {
      headerRows, bodyRows, footerRows,
      columns, loadedRowsStart, totalRowCount,
      getCellColSpan, minColumnWidth,
    } = this.props;
    const getColumnWidth = this.getColumnWidthGetter(columns, containerWidth, minColumnWidth!);

    return getCollapsedGrids({
      headerRows,
      bodyRows,
      footerRows,
      columns,
      loadedRowsStart,
      totalRowCount,
      getCellColSpan,
      viewportLeft,
      containerWidth,
      viewport,
      getRowHeight: this.getRowHeight,
      getColumnWidth,
    });
  }

  render() {
    const {
      containerComponent: Container,
      headTableComponent: HeadTable,
      footerTableComponent: FootTable,
      tableComponent: Table,
      headComponent: Head,
      bodyComponent: Body,
      footerComponent: Footer,
      tableRef,
      height,
      headerRows,
      footerRows,
      minColumnWidth,
      minWidth,
      cellComponent,
      rowComponent,
      viewport,
      scrollTop,
    } = this.props;
    const {
      headerHeight,
      bodyHeight,
      footerHeight,
    } = this.state;
    const { containerHeight } = this;

    const collapsedGrids = this.getCollapsedGrids(viewport);
    const commonProps = {
      cellComponent,
      rowComponent,
      minColumnWidth,
      minWidth,
      blockRefsHandler: this.registerBlockRef,
      rowRefsHandler: this.registerRowRef,
    };
    const sizerHeight = height === AUTO_HEIGHT ? null : height;

    return (
      <Sizer
        onSizeChange={this.handleContainerSizeChange}
        containerComponent={Container}
        style={{ height: sizerHeight }}
        onScroll={this.onScroll}
        scrollTop={scrollTop}
      >
        {
          (!!headerRows.length) && (
            <VirtualTableLayoutBlock
              {...commonProps}
              name="header"
              collapsedGrid={collapsedGrids.headerGrid}
              tableComponent={HeadTable}
              bodyComponent={Head}
            />
          )
        }
        <VirtualTableLayoutBlock
          {...commonProps}
          name="body"
          collapsedGrid={collapsedGrids.bodyGrid}
          tableComponent={Table}
          bodyComponent={Body}
          tableRef={tableRef}
          marginBottom={Math.max(0, containerHeight - headerHeight - bodyHeight - footerHeight)}
        />
        {
          (!!footerRows.length) && (
            <VirtualTableLayoutBlock
              {...commonProps}
              name="footer"
              collapsedGrid={collapsedGrids.footerGrid}
              tableComponent={FootTable}
              bodyComponent={Footer}
            />
          )
        }
      </Sizer>
    );
  }
}
