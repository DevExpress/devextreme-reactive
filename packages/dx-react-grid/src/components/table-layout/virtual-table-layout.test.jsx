import * as React from 'react';
import { findDOMNode } from 'react-dom';
import { shallow, mount } from 'enzyme';
import { setupConsole } from '@devexpress/dx-testing';
import { VirtualTableLayout } from './virtual-table-layout';
import { getCollapsedGrid } from './virtual-table-utils';

jest.mock('react-dom', () => ({
  findDOMNode: jest.fn(),
}));
jest.mock('./virtual-table-utils', () => {
  const actual = require.requireActual('./virtual-table-utils');
  jest.spyOn(actual, 'getCollapsedGrid');
  return actual;
});
jest.mock('./column-group', () => ({
  ColumnGroup: () => null,
}));
jest.mock('@devexpress/dx-react-core', () => {
  const { Component } = require.requireActual('react');
  return {
    Sizer: ({ children }) => children({ width: 400, height: 100 }),
    // eslint-disable-next-line react/prefer-stateless-function
    RefHolder: class extends Component {
      render() {
        // eslint-disable-next-line react/prop-types
        return this.props.children;
      }
    },
  };
});

const defaultProps = {
  columns: [
    { key: 'a', column: { name: 'a' } },
    { key: 'b', column: { name: 'b' } },
    { key: 'c', column: { name: 'c' } },
    { key: 'd', column: { name: 'd' } },
    { key: 'e', column: { name: 'e' } },
  ],
  minWidth: 400,
  minColumnWidth: 120,
  height: 100,
  estimatedRowHeight: 40,
  bodyRows: [
    { key: 1 },
    { key: 2 },
    { key: 3 },
    { key: 4 },
    { key: 5 },
    { key: 6 },
  ],
  containerComponent: props => <div {...props} />,
  headTableComponent: props => <table {...props} />,
  tableComponent: props => <table {...props} />,
  headComponent: props => <thead {...props} />,
  bodyComponent: props => <tbody {...props} />,
  rowComponent: () => null,
  cellComponent: () => null,
  getCellColSpan: () => 1,
};

describe('VirtualTableLayout', () => {
  let resetConsole;
  beforeEach(() => {
    resetConsole = setupConsole();
    findDOMNode.mockImplementation(() => ({
      getBoundingClientRect: () => ({
        height: defaultProps.estimatedRowHeight,
      }),
    }));
  });

  afterEach(() => {
    resetConsole();
    jest.clearAllMocks();
  });

  const simulateScroll = (tree, props) => {
    const target = {
      scrollTop: 0,
      scrollLeft: 0,
      ...props,
    };

    const eventData = {
      target,
      currentTarget: target,
    };

    tree
      .find('Sizer')
      .dive()
      .find(defaultProps.containerComponent)
      .prop('onScroll')(eventData);
    tree.update();
  };

  it('should render correct layout', () => {
    const tree = shallow((
      <VirtualTableLayout
        {...defaultProps}
        headerRows={defaultProps.bodyRows.slice(0, 1)}
      />
    ));

    expect(tree.find('Sizer').dive())
      .toMatchSnapshot();
  });

  describe('viewport', () => {
    it('should pass correct viewport at startup', () => {
      const tree = shallow((
        <VirtualTableLayout
          {...defaultProps}
          headerRows={defaultProps.bodyRows.slice(0, 1)}
        />
      ));

      tree.find('Sizer').dive();

      expect(getCollapsedGrid.mock.calls[getCollapsedGrid.mock.calls.length - 2][0])
        .toMatchObject({
          top: 0,
          left: 0,
          height: defaultProps.estimatedRowHeight,
          width: 400,
        });
      expect(getCollapsedGrid.mock.calls[getCollapsedGrid.mock.calls.length - 1][0])
        .toMatchObject({
          top: 0,
          left: 0,
          height: defaultProps.height - defaultProps.estimatedRowHeight,
          width: 400,
        });
    });

    it('should pass correct viewport at on viewport change', () => {
      const tree = shallow((
        <VirtualTableLayout
          {...defaultProps}
          headerRows={defaultProps.bodyRows.slice(0, 1)}
        />
      ));

      simulateScroll(tree, { scrollTop: 100, scrollLeft: 50 });

      tree.find('Sizer').dive();
      expect(getCollapsedGrid.mock.calls[getCollapsedGrid.mock.calls.length - 2][0])
        .toMatchObject({
          top: 0,
          left: 50,
        });
      expect(getCollapsedGrid.mock.calls[getCollapsedGrid.mock.calls.length - 1][0])
        .toMatchObject({
          top: 100,
          left: 50,
        });
    });
  });

  describe('row heights', () => {
    it('should specify correct row height at startup', () => {
      expect.hasAssertions();

      const rows = [
        { key: 1 },
        { key: 2, height: 10 },
      ];

      getCollapsedGrid
        .mockImplementationOnce((args) => {
          const { getRowHeight } = args;
          expect(getRowHeight(rows[0]))
            .toEqual(defaultProps.estimatedRowHeight);
          expect(getRowHeight(rows[1]))
            .toEqual(10);

          return require.requireActual('./virtual-table-utils').getCollapsedGrid(args);
        });

      mount((
        <VirtualTableLayout
          {...defaultProps}
          bodyRows={rows}
        />
      ));
    });

    it('should store row height when rendered', () => {
      const rows = [
        { key: 1 },
        { key: 2, height: 10 },
      ];

      findDOMNode.mockImplementation(() => ({
        getBoundingClientRect: () => ({
          height: 50,
        }),
      }));

      mount((
        <VirtualTableLayout
          {...defaultProps}
          bodyRows={rows}
        />
      ));

      const { getRowHeight } = getCollapsedGrid.mock.calls[0][0];
      expect(getRowHeight(rows[0]))
        .toEqual(50);
      expect(getRowHeight(rows[1]))
        .toEqual(50);
    });

    it('should clear row height when rows updated', () => {
      const rows = [
        { key: 11 },
        { key: 12 },
      ];

      findDOMNode.mockImplementation(() => ({
        getBoundingClientRect: () => ({
          height: 50,
        }),
      }));

      const tree = mount((
        <VirtualTableLayout
          {...defaultProps}
          bodyRows={rows}
        />
      ));
      tree.setProps({ bodyRows: [rows[0]] });

      const { getRowHeight } = getCollapsedGrid.mock.calls[0][0];
      expect(getRowHeight(rows[1]))
        .toEqual(defaultProps.estimatedRowHeight);
    });

    it('should clear row height when headerRows updated', () => {
      const rows = [
        { key: 11 },
        { key: 12 },
      ];

      findDOMNode.mockImplementation(() => ({
        getBoundingClientRect: () => ({
          height: 50,
        }),
      }));

      const tree = mount((
        <VirtualTableLayout
          {...defaultProps}
          headerRows={rows}
        />
      ));
      tree.setProps({ headerRows: [rows[0]] });

      const { getRowHeight } = getCollapsedGrid.mock.calls[0][0];
      expect(getRowHeight(rows[1]))
        .toEqual(defaultProps.estimatedRowHeight);
    });
  });
});
