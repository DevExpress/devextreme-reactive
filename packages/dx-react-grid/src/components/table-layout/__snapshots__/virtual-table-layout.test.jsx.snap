// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualTableLayout should render correct layout 1`] = `
<containerComponent
  onScroll={[Function]}
  style={
    Object {
      "height": "100px",
    }
  }
>
  <headTableComponent
    style={
      Object {
        "minWidth": "400px",
      }
    }
  >
    <ColumnGroup
      columns={
        Array [
          Object {
            "column": Object {
              "name": "a",
            },
            "key": "a",
          },
          Object {
            "column": Object {
              "name": "b",
            },
            "key": "b",
          },
          Object {
            "column": Object {
              "name": "c",
            },
            "key": "c",
          },
          Object {
            "column": Object {
              "name": "d",
            },
            "key": "d",
          },
          Object {
            "column": Object {
              "name": "e",
            },
            "key": "e",
          },
        ]
      }
    />
    <headComponent>
      <RefHolder
        key="1"
      >
        <rowComponent
          tableRow={
            Object {
              "key": 1,
            }
          }
        >
          <cellComponent
            colSpan={1}
            key="a"
            tableColumn={
              Object {
                "column": Object {
                  "name": "a",
                },
                "key": "a",
              }
            }
            tableRow={
              Object {
                "key": 1,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="b"
            tableColumn={
              Object {
                "column": Object {
                  "name": "b",
                },
                "key": "b",
              }
            }
            tableRow={
              Object {
                "key": 1,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="c"
            tableColumn={
              Object {
                "column": Object {
                  "name": "c",
                },
                "key": "c",
              }
            }
            tableRow={
              Object {
                "key": 1,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="d"
            tableColumn={
              Object {
                "column": Object {
                  "name": "d",
                },
                "key": "d",
              }
            }
            tableRow={
              Object {
                "key": 1,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="e"
            tableColumn={
              Object {
                "column": Object {
                  "name": "e",
                },
                "key": "e",
              }
            }
            tableRow={
              Object {
                "key": 1,
              }
            }
          />
        </rowComponent>
      </RefHolder>
    </headComponent>
  </headTableComponent>
  <tableComponent
    style={
      Object {
        "minWidth": "400px",
      }
    }
  >
    <ColumnGroup
      columns={
        Array [
          Object {
            "column": Object {
              "name": "a",
            },
            "key": "a",
          },
          Object {
            "column": Object {
              "name": "b",
            },
            "key": "b",
          },
          Object {
            "column": Object {
              "name": "c",
            },
            "key": "c",
          },
          Object {
            "column": Object {
              "name": "d",
            },
            "key": "d",
          },
          Object {
            "column": Object {
              "name": "e",
            },
            "key": "e",
          },
        ]
      }
    />
    <bodyComponent>
      <RefHolder
        key="1"
      >
        <rowComponent
          tableRow={
            Object {
              "key": 1,
            }
          }
        >
          <cellComponent
            colSpan={1}
            key="a"
            tableColumn={
              Object {
                "column": Object {
                  "name": "a",
                },
                "key": "a",
              }
            }
            tableRow={
              Object {
                "key": 1,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="b"
            tableColumn={
              Object {
                "column": Object {
                  "name": "b",
                },
                "key": "b",
              }
            }
            tableRow={
              Object {
                "key": 1,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="c"
            tableColumn={
              Object {
                "column": Object {
                  "name": "c",
                },
                "key": "c",
              }
            }
            tableRow={
              Object {
                "key": 1,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="d"
            tableColumn={
              Object {
                "column": Object {
                  "name": "d",
                },
                "key": "d",
              }
            }
            tableRow={
              Object {
                "key": 1,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="e"
            tableColumn={
              Object {
                "column": Object {
                  "name": "e",
                },
                "key": "e",
              }
            }
            tableRow={
              Object {
                "key": 1,
              }
            }
          />
        </rowComponent>
      </RefHolder>
      <RefHolder
        key="2"
      >
        <rowComponent
          tableRow={
            Object {
              "key": 2,
            }
          }
        >
          <cellComponent
            colSpan={1}
            key="a"
            tableColumn={
              Object {
                "column": Object {
                  "name": "a",
                },
                "key": "a",
              }
            }
            tableRow={
              Object {
                "key": 2,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="b"
            tableColumn={
              Object {
                "column": Object {
                  "name": "b",
                },
                "key": "b",
              }
            }
            tableRow={
              Object {
                "key": 2,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="c"
            tableColumn={
              Object {
                "column": Object {
                  "name": "c",
                },
                "key": "c",
              }
            }
            tableRow={
              Object {
                "key": 2,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="d"
            tableColumn={
              Object {
                "column": Object {
                  "name": "d",
                },
                "key": "d",
              }
            }
            tableRow={
              Object {
                "key": 2,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="e"
            tableColumn={
              Object {
                "column": Object {
                  "name": "e",
                },
                "key": "e",
              }
            }
            tableRow={
              Object {
                "key": 2,
              }
            }
          />
        </rowComponent>
      </RefHolder>
      <RefHolder
        key="3"
      >
        <rowComponent
          tableRow={
            Object {
              "key": 3,
            }
          }
        >
          <cellComponent
            colSpan={1}
            key="a"
            tableColumn={
              Object {
                "column": Object {
                  "name": "a",
                },
                "key": "a",
              }
            }
            tableRow={
              Object {
                "key": 3,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="b"
            tableColumn={
              Object {
                "column": Object {
                  "name": "b",
                },
                "key": "b",
              }
            }
            tableRow={
              Object {
                "key": 3,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="c"
            tableColumn={
              Object {
                "column": Object {
                  "name": "c",
                },
                "key": "c",
              }
            }
            tableRow={
              Object {
                "key": 3,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="d"
            tableColumn={
              Object {
                "column": Object {
                  "name": "d",
                },
                "key": "d",
              }
            }
            tableRow={
              Object {
                "key": 3,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="e"
            tableColumn={
              Object {
                "column": Object {
                  "name": "e",
                },
                "key": "e",
              }
            }
            tableRow={
              Object {
                "key": 3,
              }
            }
          />
        </rowComponent>
      </RefHolder>
      <RefHolder
        key="4"
      >
        <rowComponent
          tableRow={
            Object {
              "key": 4,
            }
          }
        >
          <cellComponent
            colSpan={1}
            key="a"
            tableColumn={
              Object {
                "column": Object {
                  "name": "a",
                },
                "key": "a",
              }
            }
            tableRow={
              Object {
                "key": 4,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="b"
            tableColumn={
              Object {
                "column": Object {
                  "name": "b",
                },
                "key": "b",
              }
            }
            tableRow={
              Object {
                "key": 4,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="c"
            tableColumn={
              Object {
                "column": Object {
                  "name": "c",
                },
                "key": "c",
              }
            }
            tableRow={
              Object {
                "key": 4,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="d"
            tableColumn={
              Object {
                "column": Object {
                  "name": "d",
                },
                "key": "d",
              }
            }
            tableRow={
              Object {
                "key": 4,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="e"
            tableColumn={
              Object {
                "column": Object {
                  "name": "e",
                },
                "key": "e",
              }
            }
            tableRow={
              Object {
                "key": 4,
              }
            }
          />
        </rowComponent>
      </RefHolder>
      <RefHolder
        key="5"
      >
        <rowComponent
          tableRow={
            Object {
              "key": 5,
            }
          }
        >
          <cellComponent
            colSpan={1}
            key="a"
            tableColumn={
              Object {
                "column": Object {
                  "name": "a",
                },
                "key": "a",
              }
            }
            tableRow={
              Object {
                "key": 5,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="b"
            tableColumn={
              Object {
                "column": Object {
                  "name": "b",
                },
                "key": "b",
              }
            }
            tableRow={
              Object {
                "key": 5,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="c"
            tableColumn={
              Object {
                "column": Object {
                  "name": "c",
                },
                "key": "c",
              }
            }
            tableRow={
              Object {
                "key": 5,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="d"
            tableColumn={
              Object {
                "column": Object {
                  "name": "d",
                },
                "key": "d",
              }
            }
            tableRow={
              Object {
                "key": 5,
              }
            }
          />
          <cellComponent
            colSpan={1}
            key="e"
            tableColumn={
              Object {
                "column": Object {
                  "name": "e",
                },
                "key": "e",
              }
            }
            tableRow={
              Object {
                "key": 5,
              }
            }
          />
        </rowComponent>
      </RefHolder>
      <RefHolder
        key="stub_5_5"
      >
        <rowComponent
          style={
            Object {
              "height": "40px",
            }
          }
          tableRow={
            Object {
              "height": 40,
              "key": "stub_5_5",
              "type": "stub",
            }
          }
        />
      </RefHolder>
    </bodyComponent>
  </tableComponent>
</containerComponent>
`;
